File: bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/BazelProcess.java
package org.jetbrains.bsp.bazel.bazelrunner;

import java.io.InputStream;
import java.time.Duration;
import java.util.function.Function;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.bsp.bazel.bazelrunner.outputs.AsyncOutputProcessor;
import org.jetbrains.bsp.bazel.bazelrunner.outputs.OutputCollector;
import org.jetbrains.bsp.bazel.commons.Format;
import org.jetbrains.bsp.bazel.commons.Stopwatch;
import org.jetbrains.bsp.bazel.logger.BspClientLogger;

public class BazelProcess {

  private final Process process;
  private final BspClientLogger logger;

  private static final Logger LOGGER = LogManager.getLogger(BazelProcess.class);

  BazelProcess(Process process, BspClientLogger logger) {
    this.process = process;
    this.logger = logger;
  }

  public BazelProcessResult waitAndGetResult() {
    try {
      var stdoutCollector = new OutputCollector();
      var stderrCollector = new OutputCollector();
      var outputProcessor = new AsyncOutputProcessor();
      var stopwatch = Stopwatch.start();
      outputProcessor.start(
          process.getInputStream(), stdoutCollector, logger::message, LOGGER::info);
      outputProcessor.start(process.getErrorStream(), stderrCollector, logger::error, LOGGER::info);
      var exitCode = process.waitFor();
      outputProcessor.shutdown();
      var duration = stopwatch.stop();
      logCompletion(exitCode, duration);
      return new BazelProcessResult(stdoutCollector, stderrCollector, exitCode);
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
  }

  public <T> T processBinaryOutput(Function<InputStream, T> func) {
    try {
      var stopwatch = Stopwatch.start();
      var result = func.apply(process.getInputStream());
      var exitCode = process.waitFor();
      var duration = stopwatch.stop();
      logCompletion(exitCode, duration);
      return result;
    } catch (InterruptedException e) {
      throw new RuntimeException(e);
    }
  }

  private void logCompletion(int exitCode, Duration duration) {
    logger.message("Command completed in %s (exit code %d)", Format.duration(duration), exitCode);
  }
}

File: bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/BazelProcess.kt
package org.jetbrains.bsp.bazel.bazelrunner

import java.time.Duration
import org.apache.logging.log4j.LogManager
import org.eclipse.lsp4j.jsonrpc.CancelChecker
import org.jetbrains.bsp.bazel.bazelrunner.outputs.AsyncOutputProcessor
import org.jetbrains.bsp.bazel.bazelrunner.outputs.OutputProcessor
import org.jetbrains.bsp.bazel.bazelrunner.outputs.SyncOutputProcessor
import org.jetbrains.bsp.bazel.commons.Format
import org.jetbrains.bsp.bazel.commons.Stopwatch
import org.jetbrains.bsp.bazel.logger.BspClientLogger

class BazelProcess internal constructor(
    private val process: Process,
    private val logger: BspClientLogger,
    private val originId: String?
) {

    fun waitAndGetResult(cancelChecker: CancelChecker, ensureAllOutputRead: Boolean = false): BazelProcessResult {
        val stopwatch = Stopwatch.start()
        val outputProcessor: OutputProcessor =
          if (ensureAllOutputRead) SyncOutputProcessor(process, logger::message, LOGGER::info)
          else AsyncOutputProcessor(process, logger::message, LOGGER::info)

        val exitCode = outputProcessor.waitForExit(cancelChecker)
        val duration = stopwatch.stop()
        logCompletion(exitCode, duration)
        return BazelProcessResult(outputProcessor.stdoutCollector, outputProcessor.stderrCollector, exitCode)
    }

    private fun logCompletion(exitCode: Int, duration: Duration) {
        logger.withOriginId(originId)
            .message("Command completed in %s (exit code %d)", Format.duration(duration), exitCode)
    }

    companion object {
        private val LOGGER = LogManager.getLogger(BazelProcess::class.java)
    }
}
